From 6409e7244aa13c29265a23cc1fd6daa5676792dd Mon Sep 17 00:00:00 2001
From: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Date: Fri, 8 Nov 2024 15:06:44 +0100
Subject: [PATCH 1435/1695] drivers: stm32_rif: Check availability before
 taking a semaphore

Check availability before taking the semaphore to avoid IAC.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git d0dcec945612ab54ed90748bac9f6f2cd91c49fc drivers: stm32_rif: Check availability before taking a semaphore]
Signed-off-by: Gabriel Fernandez <gabriel.fernandez@foss.st.com>
Change-Id: I598043334130428c39baabf3b8ede91dbb5a9baa
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/417412
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Etienne CARRIERE <etienne.carriere@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/firewall/stm32_rif.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/core/drivers/firewall/stm32_rif.c b/core/drivers/firewall/stm32_rif.c
index 30dcf872a..18b31266c 100644
--- a/core/drivers/firewall/stm32_rif.c
+++ b/core/drivers/firewall/stm32_rif.c
@@ -100,8 +100,9 @@ TEE_Result stm32_rif_acquire_semaphore(vaddr_t addr, unsigned int nb_cid_supp)
 {
 	uint32_t scid_mask = get_scid_mask(nb_cid_supp);
 
-	/* Take the semaphore */
-	io_setbits32(addr, _SEMCR_MUTEX);
+	/* Check availability before taking the semaphore to avoid IAC */
+	if (stm32_rif_semaphore_is_available(addr))
+		io_setbits32(addr, _SEMCR_MUTEX);
 
 	/* Check that the Cortex-A has the semaphore */
 	if (stm32_rif_semaphore_is_available(addr) ||
-- 
2.47.2

